stellargraph/layer/deep_graph_infomax.py

```diff
@@ -0,0 +1,137 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from . import GCN, GAT, APPNP, PPNP
+from tensorflow.keras.layers import Input, Dense, Lambda, Layer
+import tensorflow as tf
+from tensorflow.keras import backend as K
+from ..core.experimental import experimental
+
+
+__all__ = [
+    "fullbatch_infomax_node_model",
+    "fullbatch_infomax_embedding_model",
+]
+
+_NODE_FEATS = "GCN_INFO_MAX_NODE_FEATURES"
+
+
+class Discriminator(Layer):
+    """
+    This Layer computes the Discriminator function for Deep Graph Infomax (https://arxiv.org/pdf/1809.10341.pdf).
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def build(self, input_shapes):
+
+        self.kernel = self.add_weight(
+            shape=(input_shapes[0][1], input_shapes[0][1]),
+            initializer="glorot_uniform",
+            name="kernel",
+            regularizer=None,
+            constraint=None,
+        )
+        self.built = True
+
+    def call(self, inputs):
+        """
+        Applies the layer to the inputs.
+
+        Args:
+            inputs: a list of tensors with shapes [(N, F), (F,)] containing the node features and summary feature
+                vector.
+        """
+        features, summary = inputs
+
+        score = tf.linalg.matvec(tf.linalg.matmul(features, self.kernel), summary)
+
+        return score
+
+
+@experimental(reason="lack of unit tests", issues=[1003])
+def fullbatch_infomax_node_model(base_model):
+    """
+    A function to create the the inputs and outputs for a Deep Graph Infomax model for unsupervised training.
+
+    Args:
+        base_model: the base stellargraph model class
+
+    Returns:
+        input and output layers for use with a keras model
+    """
+
+    if not isinstance(base_model, (GCN, GAT, APPNP, PPNP)):
+        raise TypeError(
+            f"base_model: expected GCN, GAT, APPNP or PPNP found {type(base_model).__name__}"
+        )
+
+    # Inputs for features
+    x_t = Input(batch_shape=(1, base_model.n_nodes, base_model.n_features))
+    # Inputs for shuffled features
+    x_corr = Input(batch_shape=(1, base_model.n_nodes, base_model.n_features))
```

Comment:
Here we assume that the source and corrupted graphs are the same size. This will work on the single input graph transductive settings, e.g., Cora, and for the corruption function that only shuffles the features and keeps the adjacency matrix the same. This is the simplest setting for using this method and this should work. 

We should plan ahead for the multi-graph case where there is no guarantee that the graphs will have the same size. Our full batch GCN and GAT methods (not sure about APPNP and PPNP) cannot handle input graphs of different size. 

I don't want to suggest that you update the code now, but maybe create a new ticket to keep track of this issue. We will come across this problem sooner than later.